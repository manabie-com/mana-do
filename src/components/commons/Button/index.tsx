import React, { forwardRef } from "react";
import cn from "classnames";
import "./style.css";

export enum ButtonSizes {
  lg = "lg",
  md = "md",
  xs = "xs",
}

export enum ButtonColors {
  primary = "primary",
  danger = "danger",
}

interface ButtonProps {
  disabled?: boolean;
  fullWidth?: boolean;
  type?: "submit" | "button" | "reset";
  onClick?: (event: React.SyntheticEvent) => void;
  color?: ButtonColors;
  className?: string;
  children?: React.ReactNode;
}

const defaultProps: ButtonProps = {
  disabled: false,
  fullWidth: true,
  type: "button",
  color: ButtonColors.primary,
};

const Button = forwardRef((props: ButtonProps, ref: any) => {
  const { color, className, children, disabled, fullWidth, type, ...rest } = {
    ...defaultProps,
    ...props,
  };

  const classOfComponent = cn(
    className,
    "btn-custom",
    `btn-custom-${color}`,
    disabled && "btn-custom-disabled"
  );

  return (
    <button
      className={classOfComponent}
      ref={ref}
      disabled={disabled}
      type={type}
      {...rest}
    >
      {children}
    </button>
  );
});

export default Button;
