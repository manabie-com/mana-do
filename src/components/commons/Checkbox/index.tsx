import React, { forwardRef, useRef } from "react";
import cn from "classnames";
import CheckboxChecked from "./icon/CheckboxChecked";
import CheckboxUnChecked from "./icon/CheckboxUnChecked";
import Input from "../Input";
import "./style.css";

interface Props {
  className?: string;
  iconRef?: React.Ref<unknown>;
  disabled?: boolean;
  children?: React.ReactNode;
  checked?: boolean;
  value?: string;
  name?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const defaultProps: Props = {
  disabled: false,
  checked: false,
};

const Checkbox = forwardRef((props: Props, ref: any) => {
  const {
    className,
    iconRef,
    name,
    disabled,
    checked,
    value,
    children,
    onChange,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  };

  const classOfRoot = cn(
    "checkbox-custom",
    className,
    disabled && "checkbox-custom-disabled",
    checked && "checkbox-custom-checked"
  );

  const refOfInput = useRef();

  return (
    <label
      data-testid="checkbox-custom"
      className={classOfRoot}
      ref={ref}
      {...rest}
    >
      {checked ? (
        <CheckboxChecked ref={iconRef} />
      ) : (
        <CheckboxUnChecked ref={iconRef} />
      )}

      {children}
      <Input
        readOnly
        type="checkbox"
        ref={refOfInput}
        name={name}
        value={value}
        className="checkbox-custom-input"
        checked={checked}
        onChange={onChange}
      />
    </label>
  );
});

export default Checkbox;
