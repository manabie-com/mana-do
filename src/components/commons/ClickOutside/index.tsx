import React, { useRef, useEffect } from "react";

/**
 * Hook that alerts clicks outside of the passed ref
 */
function useClickOutsideComponent(ref: any, onClickOutside: () => void) {
  useEffect(() => {
    /**
     * Alert if clicked on outside of element
     */
    function handleClickOutside(event: any) {
      if (ref.current && !ref.current.contains(event.target)) {
        onClickOutside();
      }
    }
    // Bind the event listener
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [ref, onClickOutside]);
}

/**
 * Component that alerts if you click outside of it
 */
interface Props {
  onClickOutside?: () => void;
  children?: React.ReactNode;
}

function ClickOutsideComponent(props: Props) {
  const { onClickOutside = () => null, children } = props;

  const wrapperRef = useRef(null);
  useClickOutsideComponent(wrapperRef, onClickOutside);

  return <div ref={wrapperRef}>{children}</div>;
}

export default ClickOutsideComponent;
