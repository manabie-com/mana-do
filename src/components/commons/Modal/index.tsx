import React, { forwardRef } from "react";
import cn from "classnames";
import "./style.css";

interface Props {
  show?: boolean;
  width?: string;
  children?: React.ReactNode;
  title?: string;
  onClose?: () => void;
  className?: string;
}

const defaultProps: Props = {
  show: false,
  width: "30%",
  onClose: () => null,
};

const Modal = forwardRef((props: Props, ref: any) => {
  const { onClose, title, width, show, className, children, ...rest } = {
    ...defaultProps,
    ...props,
  };
  const showHideClassName = show
    ? "modal-custom display-block"
    : "modal-custom display-none";
  return (
    <div className={cn(showHideClassName, className)} {...rest}>
      <section style={{ width: width }} className="modal-main">
        <div className="header-modal mb40">
          <span className="title">{title}</span>
          <span className="icon-close" onClick={onClose}>
            <img width="30" src="images/cancel.png" alt="icon-close" />
          </span>
        </div>
        <div className="content-modal mt30">{children}</div>
      </section>
    </div>
  );
});

export default Modal;
