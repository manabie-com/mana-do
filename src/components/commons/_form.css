form label {
  width: 100%;
  padding-top: 10px;
  margin-top: 10px;
  display: flex;
  justify-content: center;
}
form label h2 {
  margin-right: 1em;
  display: flex;
  align-items: center;
  text-align: center;
}
form label input {
  min-width: 325px;
  height: 50px;
  border: solid 1px;
  border-radius: 25px;
  padding-left: 25px;
  font-size: 18px;
  font-weight: bold;
}
form label input:hover {
  border-color: yellow;
}


@media screen and (max-width: 480px) {
  form {
    font-size: 12px;
  }
  form label input {
    min-width: 200px;
    height: 30px;
    padding-left: 15px;
  }
  form .actions button {
    min-width: 150px;
    width: auto;
  }
}