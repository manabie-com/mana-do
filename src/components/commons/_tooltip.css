.use-tooltip {
  position: relative;
  display: inline;
  text-align: left;
  width: 100%;
}
  
.use-tooltip:hover .tooltip {
  display: block;
  z-index: 3;
  transform: translateX(-50%);
  display: -webkit-box;
  -webkit-line-clamp: 5;
  word-break: break-word;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0.5em;
  font-size: 16px;
  line-height: 24px;
  color: steelblue;
  font-weight: 500;
}

.use-tooltip:hover .tooltip.top {
  position: fixed;
  top: -20px;
  left: 50%;
  min-width: 80vw;
}

.use-tooltip:hover .tooltip.center {
  text-align: center;
}

.use-tooltip:hover .tooltip.left {
  text-align: left;
}

.use-tooltip:hover .tooltip.right {
  text-align: right;
}

.use-tooltip:hover .tooltip.current {
  
}
  
.tooltip {
  display: none;
  color: white;
  background-color: rgb(240, 247, 181, 0.7);
  margin: 2em;
  max-width: 80vw;
  border: 1px solid rgba(0, 0, 0, 0.4);
  border-radius: 5px;
  box-shadow: 4px 2px rgba(0,0,0, 0.4);
}

.typewriter {
}

.typewriter .tooltip-text {
  text-align: right;
  white-space: nowrap;

  /* animation properties */
  -moz-transform: translateX(-100%);
  -webkit-transform: translateX(-100%);
  transform: translateX(-100%);
  
  -moz-animation: running-text 20s linear infinite;
  -webkit-animation: running-text 20s linear infinite;
  animation: running-text 20s linear infinite;
}

/* for Firefox */
@-moz-keyframes running-text {
  from { -moz-transform: translateX(-100%); }
  to { -moz-transform: translateX(100%); }
}

/* for Chrome */
@-webkit-keyframes running-text {
  from { -webkit-transform: translateX(-100%); }
  to { -webkit-transform: translateX(100%); }
}

@keyframes running-text {
  from {
    -moz-transform: translateX(-100%);
    -webkit-transform: translateX(-100%);
    transform: translateX(-100%);
  }
  to {
    -moz-transform: translateX(100%);
    -webkit-transform: translateX(100%);
    transform: translateX(100%);
  }
}