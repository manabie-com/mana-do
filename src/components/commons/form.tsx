import React, { useCallback } from 'react'

export type FormProps = {
  model: any,
  formData: any,
  onSubmit: Function,
  onUpdateField: Function,
  submitText?: string
}

const Form = (props: FormProps) => {
  const { model, formData, onSubmit
    , onUpdateField, submitText
  } = props
  const fields = Object.keys(model).map((key) => {
    return model[key]
  })

  const handleSubmit = useCallback((e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    onSubmit()
  }, [onSubmit])

  const handleSubmitClick = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault()
    onSubmit()
  }, [onSubmit])

  return (
    <form onSubmit={handleSubmit}>
      {fields.map((field) => {
        const { component, ...nestedProps } = field
        return component({ ...nestedProps, ...formData[field.name], onUpdateField })
      })}
      <div className='actions'>
        <button type='submit' onClick={handleSubmitClick}>
          {submitText}
        </button>
      </div>
    </form>
  )
}

export default Form