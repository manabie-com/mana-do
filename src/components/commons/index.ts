import Input from "./Input";
import Button, { ButtonColors } from "./Button";
import PrivateRoute from "./PrivateRoute";
import Paper from "./Paper";
import Checkbox from "./Checkbox";
import Modal from "./Modal";
import ModalConfirm from "./ModalConfirm";
import ClickOutsideComponent from "./ClickOutside";
export {
  PrivateRoute,
  Input,
  Button,
  ButtonColors,
  Paper,
  Checkbox,
  Modal,
  ModalConfirm,
  ClickOutsideComponent,
};
