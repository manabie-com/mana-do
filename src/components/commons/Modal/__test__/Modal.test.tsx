import React from "react";
import ReactDOM from "react-dom";
import { render, cleanup } from "@testing-library/react";
import { Modal } from "components/commons";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";

const propsTest = {
  show: true,
  width: "300px",
  title: "Title modal",
  onClose: jest.fn(),
};

let container: HTMLElement;

beforeEach(() => {
  container = render(<Modal {...propsTest} />).container;
});

afterEach(cleanup);

test("Modal: render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Modal {...propsTest} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
test("Modal: props of modal", () => {
  const modalEl = container.querySelector(".modal-custom") as HTMLDivElement;
  const modalMain = container.querySelector(".modal-main") as HTMLDivElement;
  expect(modalEl?.className).toContain("display-block");
  expect(modalMain?.style.width).toEqual(propsTest.width);
});
test("Modal: onClose call when click", () => {
  const iconCloseEl = container.querySelector(".icon-close") as HTMLSpanElement;
  userEvent.click(iconCloseEl);
  expect(propsTest.onClose).toBeCalled();
});
test("Modal: match snapshot", () => {
  const tree = renderer.create(<Modal {...propsTest} />).toJSON();
  expect(tree).toMatchSnapshot();
});
