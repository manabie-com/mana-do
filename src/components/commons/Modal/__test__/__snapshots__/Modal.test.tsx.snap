// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal: match snapshot 1`] = `
<div
  className="modal-custom display-block"
>
  <section
    className="modal-main"
    style={
      Object {
        "width": "300px",
      }
    }
  >
    <div
      className="header-modal mb40"
    >
      <span
        className="title"
      >
        Title modal
      </span>
      <span
        className="icon-close"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [
                Class {
                  "_dispatchInstances": null,
                  "_dispatchListeners": null,
                  "_targetInst": null,
                  "altKey": null,
                  "bubbles": null,
                  "button": null,
                  "buttons": null,
                  "cancelable": null,
                  "clientX": null,
                  "clientY": null,
                  "ctrlKey": null,
                  "currentTarget": [Function],
                  "defaultPrevented": null,
                  "detail": null,
                  "dispatchConfig": null,
                  "eventPhase": null,
                  "getModifierState": [Function],
                  "isDefaultPrevented": [Function],
                  "isPropagationStopped": [Function],
                  "isTrusted": null,
                  "metaKey": null,
                  "movementX": [Function],
                  "movementY": [Function],
                  "nativeEvent": null,
                  "pageX": null,
                  "pageY": null,
                  "relatedTarget": [Function],
                  "screenX": null,
                  "screenY": null,
                  "shiftKey": null,
                  "target": null,
                  "timeStamp": [Function],
                  "type": null,
                  "view": null,
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
      >
        <img
          alt="icon-close"
          src="images/cancel.png"
          width="30"
        />
      </span>
    </div>
    <div
      className="content-modal mt30"
    />
  </section>
</div>
`;
